fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/ExplainSelfContainedSymbolMap.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * ExplainSelfContainedSymbolMap.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades;
+
+import com.apple.foundationdb.record.query.plan.cascades.values.Value;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Formatting for {@link Value#explain}.
+ */
+public class ExplainSelfContainedSymbolMap implements ExplainSymbolMap {
+    @Nonnull private final AtomicInteger quantifierNumber;
+    @Nonnull private final BiMap<CorrelationIdentifier, String> aliasToFormattingNameMap;
+
+    public ExplainSelfContainedSymbolMap() {
+        this.quantifierNumber = new AtomicInteger(0);
```

Comment:
I might be a little confused as to how scopes work, but if you are using this explain symbol map in a scoped situation, is there anything stopping different maps in different scopes from choosing the same value for different quantifiers? And is that an issue?

Reply:
Scopes are needed for e.g. for the current element "_". Different scopes may contain the same mappings, but just as in other languages they get overshadowed. In reality we don't have that problem as the planner does not choose quantifiers in a way that they overshadow each other ever. And if they did then they would also overshadow each other in the mapped symbol (they would map to the same symbol).